### PR TITLE
Bluetooth: Mesh: test secure network beacon interval

### PR DESCRIPTION
**Pull request by m-alperen-sener**
the minimum time between two consecutive secure mesh beacons
should be approximately 10s

Signed-off-by: Alperen Sener <alperen.sener@nordicsemi.no>

### CODE REVIEW

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 1]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
 	PASS();
 }
 
+static void test_tx_beacon_init(void)
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:238]: No necessity in `test_tx_beacon_init` and `test_rx_beacon_init`. Just reuse already existing `test_tx_init` and `test_rx_init`. Incoming waiting time in `bt_mesh_test_cfg_set` is not related to beacon timing.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 2]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+	delay_time = get_device_nbr() * 100;
+	k_sleep(K_MSEC(delay_time));
+
+	bt_mesh_test_setup();
+
+	sub = bt_mesh_subnet_find(NULL, NULL);
+
+	sent_beacons = 0;
+	sent_beacon_time = 0;
+	while (true) {
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:265]: It is better to not do an infinite loop in tests. If functionality is broken then the test CI plan will get stuck forever. Just use one more semaphore or any other suitable Zephyr service.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 3]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+
+static void test_tx_beacon_secure(void)
+{
+	uint32_t delay_time;
+	uint32_t sent_beacons;
+	uint32_t sent_beacon_time;
+	struct bt_mesh_subnet *sub;
+	
+	/* shift beaconing time line to avoid boundary cases. */
+	delay_time = get_device_nbr() * 100;
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:256]: Hm. Why do you need it? My assumption is that beacon observation logic should manage the initial beacon storm, is that?

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 4]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+	ASSERT_EQUAL(BT_GAP_ADV_TYPE_ADV_NONCONN_IND, info->adv_type);
+
+	length = net_buf_simple_pull_u8(buf);
+	ASSERT_EQUAL(buf->len, length);
+	ASSERT_EQUAL(BT_DATA_MESH_BEACON, net_buf_simple_pull_u8(buf));
+	ASSERT_EQUAL(0x01, net_buf_simple_pull_u8(buf));
+
+	sub = bt_mesh_subnet_find(NULL, NULL);
+
+	memcpy(payload, buf->data, sizeof(payload));
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:308]:
```suggestion
	memcpy(payload, net_buf_simple_pull_mem(buf, sizeof(payload)), sizeof(payload));
```

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 5]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+	net_buf_simple_pull(buf, 13);
+
+	if (bt_mesh_aes_cmac_one(sub->keys->beacon, payload, sizeof(payload), mac)) {
+		return;
+	}
+
+	memcpy(&auth, mac, sizeof(auth));
+	sub_auth = (uint64_t *)sub->auth;
+
+	ASSERT_EQUAL(auth, *sub_auth);
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:318]:
```suggestion
	ASSERT_EQUAL(*(uint64_t *)mac, *(uint64_t *)sub->auth);
```

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 6]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+		obs_min_interval = interval;
+	}
+
+	total_beacons++;
+	if (total_beacons >= (EXPECTED_BEACON_CNT)) {
+		k_sem_give(&observer_sem);
+	}
+}
+
+static void scan_timeout_cb(void)
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:332]: Why do you need empty function?

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 7]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+static struct bt_le_scan_cb scan_callbacks = {
+	.recv = scan_recv_cb,
+	.timeout = scan_timeout_cb,
+};
+
+static void test_rx_beacon_secure(void)
+{
+	k_sem_init(&observer_sem, 0, 1);
+	/*initial estimated interval value*/
+	obs_min_interval = 10000;
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:346]: If there were no beacons then the test will pass. I guess the initial value of `obs_min_interval` should be wrong in advance.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/beacon/minimun_beacon_interval.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+# Copyright 2021 Nordic Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
+
+RunTest mesh_beacon_time_limit \
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/beacon/minimun_beacon_interval.sh:7]: Please add minor description of test scenario here

**Review by alxelax — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c (hunk 3 above)
Comment (m-alperen-sener) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:256]: Observation logic is OK but, nodes starts perfectly synchronous  and stays that way indefinitely.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c (hunk 7 above)
Comment (m-alperen-sener) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:346]: good point, I should have chose a value just below the minimum.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c (hunk 2 above)
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:265]: If I don't miss anything this while-loop could be replaced by something like this:
```
k_sleep(K_SECONDS(BEACON_OBSERVE_INTERVAL));
ASSERT_TRUE(sub->beacon_sent > 0);
PASS();
```

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 9]
```diff
@@ -244,6 +372,9 @@ static const struct bst_test_instance test_beacon[] = {
 
 	TEST_CASE(rx, on_iv_update,   "Beacon: receive with IV update flag"),
 	TEST_CASE(rx, on_key_refresh,  "Beacon: receive with key refresh flag"),
+
+	TEST_CASE(tx_beacon, secure, "Beacon: send secure mesh beacons"),
```
Comment (PavelVPV) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:376]: This sounds a bit generic. I would rephrase the test name to emphasize that its purpose is to check beacon interval.

**Review by PavelVPV — CHANGES_REQUESTED** [review]
Looks ok. Just few minor comments.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c [hunk 10]
```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
...
+	uint8_t length;
+	uint8_t payload[13];
+	uint64_t *sub_auth;
+	uint64_t auth;
+	uint8_t mac[16];
+	uint32_t now;
+	uint32_t interval;
+
+	now = k_uptime_get_32();
+	interval = now - last_beacon_time;
```
[the remark this answers is not in the document]
Comment (m-alperen-sener) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:297]: subnet uses 32bit timestamp, and delta has no 32bit version. But it could be used though.
